### PR TITLE
Adsk Contrib - Add new ACES builtin transforms

### PR DESCRIPTION
**Pull request by doug-walker**
Add Built-in Transforms requested by the ACES Config working group.
- View Transforms needed for ACES Output Transforms
- Display color spaces needed for ACES Output Transforms
- Sony Venice IDTs
- LMT-Blue light fix

Signed-off-by: Doug Walker <Doug.Walker@autodesk.com>

### CODE REVIEW

File: src/OpenColorIO/transforms/builtins/ACES.cpp [hunk 1]
```diff
@@ -115,6 +137,214 @@ void GenerateOps(OpRcPtrVec & ops)
 
 }  // namespace ADX_to_ACES
 
+namespace ACES_OUTPUT
+{
+
+void Generate_ACES_to_APO_ops(OpRcPtrVec & ops)
```
[the remark this answers is not in the document]
Comment (doug-walker) [reply src/OpenColorIO/transforms/builtins/ACES.cpp:143]: Thanks, good point.  Will change that.

File: src/OpenColorIO/transforms/builtins/ACES.cpp [hunk 2]
```diff
@@ -115,6 +137,214 @@ void GenerateOps(OpRcPtrVec & ops)
...
+        auto curve = GradingBSplineCurve::Create({
+                {-5.26017743f, -4.f},
+                {-3.75502745f, -3.57868829f},
+                {-2.24987747f, -1.82131329f},
+                {-0.74472749f,  0.68124124f},
+                { 1.06145248f,  2.87457742f},
+                { 2.86763245f,  3.83406206f},
+                { 4.67381243f,  4.f}
+            });
+        float slopes[] = { 0.f,  0.55982688f,  1.77532247f,  1.55f,  0.8787017f,  0.18374463f,  0.f };
```
Comment (hodoulp) [inline src/OpenColorIO/transforms/builtins/ACES.cpp:189]: [Minor] Don't forget to put `static constexpr` or `static const` when possible.

File: src/OpenColorIO/transforms/builtins/ACES.cpp [hunk 3]
```diff
@@ -115,6 +137,214 @@ void GenerateOps(OpRcPtrVec & ops)
...
+        auto gc = std::make_shared<GradingRGBCurveOpData>(GRADING_LOG, z, z, z, m);
+
+        CreateGradingRGBCurveOp(ops, gc, TRANSFORM_DIR_FORWARD);
+    }
+    // Undo the logarithm.
+    CreateLogOp(ops, 10., TRANSFORM_DIR_INVERSE);
+
+    // Apply Cinema White/Black correction.
+    {
+        const double Y_MIN = 0.0001;
```
Comment (hodoulp) [inline src/OpenColorIO/transforms/builtins/ACES.cpp:303]: [Minor] Align to the '='.

File: tests/cpu/CMakeLists.txt [hunk 4]
```diff
@@ -114,6 +114,7 @@ if(OCIO_ADD_EXTRA_BUILTINS)
     set(SOURCES_BUILTINS
         transforms/builtins/ArriCameras.cpp
         transforms/builtins/CanonCameras.cpp
+		transforms/builtins/Displays.cpp
```
Comment (zachlewis) [inline tests/cpu/CMakeLists.txt:117]:
```suggestion
	transforms/builtins/Displays.cpp
```
"Sorry, but i'm a bit of a stickler for paperwork..."

File: src/OpenColorIO/transforms/builtins/ColorMatrixHelpers.h [hunk 5]
```diff
@@ -153,6 +168,16 @@ MatrixOpData::MatrixArrayPtr build_conversion_matrix(const Primaries & src_prims
                                                      const MatrixOpData::Offsets & src_wht_XYZ,
                                                      const MatrixOpData::Offsets & dst_wht_XYZ,
                                                      AdaptationMethod method);
+
+// Build a conversion matrix to CIE XYZ D65 to the destination primaries.
```
Comment (hodoulp) [inline src/OpenColorIO/transforms/builtins/ColorMatrixHelpers.h:172]: '... to the destination ...` -> `... from the source ...` ?

File: src/OpenColorIO/transforms/builtins/Displays.cpp [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+static constexpr double c3 = 32. * 2392. / 4096.;
+static constexpr double c1 = c3 - c2 + 1.;
+
+void GeneratePQToLinearOps(OpRcPtrVec & ops)
+{
+    auto GenerateLutValues = [](double input) -> float
+    {
+        const double N = std::max(0., input);
+        const double x = std::pow(N, 1. / m2);
+        double L = std::pow( std::max(0., x - c1) / (c2 - c3 * x), 1. / m1 );
```
Comment (hodoulp) [inline src/OpenColorIO/transforms/builtins/Displays.cpp:39]: [Minor] The variable could be const.

File: tests/cpu/CMakeLists.txt (hunk 4 above)
Comment (hodoulp) [inline tests/cpu/CMakeLists.txt:117]: [Minor] Tabulation issue.

File: src/OpenColorIO/transforms/builtins/Displays.cpp (hunk 6 above)
Comment (doug-walker) [reply src/OpenColorIO/transforms/builtins/Displays.cpp:39]: It gets modified.

File: src/OpenColorIO/ops/matrix/MatrixOpData.cpp [hunk 7]
```diff
@@ -750,7 +750,8 @@ void MatrixOpData::cleanUp(double offsetScale)
     // args.  In any case, the tolerance is small enough to pick up anything
     // that would be significant in the context of color management.
     const double scale = max_val > 1e-4 ? max_val : 1e-4;
-    const double abs_tol = scale * 1e-6;
+//    const double abs_tol = scale * 1e-6;
```
Comment (doug-walker) [inline src/OpenColorIO/ops/matrix/MatrixOpData.cpp:753]: Will remove.

File: tests/cpu/transforms/BuiltinTransform_tests.cpp [hunk 8]
```diff
@@ -343,6 +343,7 @@ void ValidateBuiltinTransform(const char * style, const Values & in, const Value
     OCIO::ConstCPUProcessorRcPtr cpu;
     // Use lossless mode for these tests (e.g. FAST_LOG_EXP_POW limits to about 4 sig. digits).
     OCIO_CHECK_NO_THROW_FROM(cpu = proc->getOptimizedCPUProcessor(OCIO::OPTIMIZATION_LOSSLESS), lineNo);
+//     OCIO_CHECK_NO_THROW_FROM(cpu = proc->getOptimizedCPUProcessor(OCIO::OPTIMIZATION_NONE), lineNo);
```
Comment (doug-walker) [inline tests/cpu/transforms/BuiltinTransform_tests.cpp:346]: Will remove.

File: src/OpenColorIO/transforms/builtins/ACES.cpp [hunk 9]
```diff
@@ -598,26 +843,190 @@ void RegisterAll(BuiltinTransformRegistryImpl & registry) noexcept
...
+            CreateScaleOp(ops, scale4, TRANSFORM_DIR_FORWARD);
+
+            AP1_to_CIE_XYZ_D65::GenerateOps(ops);
+
+            MatrixOpData::MatrixArrayPtr matrix2
+                = build_vonkries_adapt(WHITEPOINT::DCI_XYZ, WHITEPOINT::D65_XYZ, ADAPTATION_BRADFORD);
+            CreateMatrixOp(ops, matrix2, TRANSFORM_DIR_FORWARD);
+        };
+
+        registry.addBuiltin("ACES-OUTPUT--ACES_to_CIE-XYZ_SDR-CINEMA-D65sim-DCI_1.1",
```
Comment (hodoulp) [inline src/OpenColorIO/transforms/builtins/ACES.cpp:908]: As the builtin names end up to be public, I have two questions:

- Why did you change `ACES2065-1` to `ACES`?
I found it less readable when present in a config file.

- What is the meaning of `--`?   
If that's a way to add some `categories` (or `families`) to the builtin transforms a solution is to explicitly add them i.e. `registry.addBuiltin(tag, family tag, explanation, functor)` and use the family 'style' (i.e. `/`) in the config file.

Comment (michdolan) [reply src/OpenColorIO/transforms/builtins/ACES.cpp:908]: I had a similar thought about the `--`.

**Comment by michdolan** [comment]
Only other comment from my end is if you think anyone in TSC will take issue with use of `CINEMA`/`VIDEO` in builtin transform names. It has been present in the spreadsheet for some time and nobody commented, but may be worth asking that question at the TSC meeting. I don't personally have an issue with it, and understand why it's there, but curious what others think. I know that pushes back releasing this by a day, but IMO we're on track if we get v2.0.0 out by end of month, so another day or two is OK.

File: src/OpenColorIO/transforms/builtins/ACES.cpp [hunk 10]
```diff
@@ -598,26 +843,190 @@ void RegisterAll(BuiltinTransformRegistryImpl & registry) noexcept
...
+            ACES_OUTPUT::Generate_RRT_preamble_ops(ops);
+
+            ACES_OUTPUT::Generate_hdr_tonecurve_ops(ops, 1000.);
+
+            ACES_OUTPUT::Generate_hdr_primary_clamp_ops(ops, REC2020::primaries);
+
+            ACES_OUTPUT::Generate_nit_normalization_ops(ops, 1000.);
+        };
+
+        registry.addBuiltin("ACES-OUTPUT--ACES_to_CIE-XYZ_HDR-VIDEO-1000nit-15nit-REC2020lim_1.1",
```
Comment (hodoulp) [inline src/OpenColorIO/transforms/builtins/ACES.cpp:925]: Personally, I find that some new builtin names are less and less readable. A suggestion:
`ACES-OUTPUT / ACES2065_1__to__CIE-XYZ_HDR-VIDEO-1000nit-15nit-REC2020lim_1.1`. But the last part is still 'dense'!

And the explanation does not seem to contain all the information (as in the builtin name)

**Comment by michdolan** [comment]
Awesome work @doug-walker ! Thanks for the time and energy you have put into this.

File: src/OpenColorIO/transforms/builtins/ACES.cpp [hunk 11]
```diff
@@ -598,26 +843,190 @@ void RegisterAll(BuiltinTransformRegistryImpl & registry) noexcept
...
 
-            ACES_OUTPUT::Generate_primary_clamp_ops(ops);
+            ACES_OUTPUT::Generate_hdr_primary_clamp_ops(ops, P3_D65::primaries);
 
             ACES_OUTPUT::Generate_nit_normalization_ops(ops, 1000.);
         };
 
-        registry.addBuiltin("ACES-OUTPUT--ACES2065-1_to_CIE-XYZ_HDR-VIDEO-1000nits_1.1", 
+        registry.addBuiltin("ACES-OUTPUT--ACES_to_CIE-XYZ_HDR-VIDEO-1000nit-15nit-P3lim_1.1", 
                             "Component of ACES Output Transforms for 1000 nit HDR D65 video",
```
Comment (hodoulp) [inline src/OpenColorIO/transforms/builtins/ACES.cpp:943]: Same explanation than the one for the builtin transform at line 926.

**Review by michdolan — APPROVED** [review]

**Review by hodoulp — APPROVED** [review]

**Pull request merged**